components/eam/src/physics/cam/hetfrz_classnuc_cam.F90

```diff
@@ -876,43 +895,211 @@ end subroutine hetfrz_classnuc_cam_save_cbaero
-   real(r8), intent(out) :: total_cloudborne_aer_num(3) ! #/cm^3
-   real(r8), intent(out) :: coated_aer_num(3)           ! #/cm^3 
-   real(r8), intent(out) :: uncoated_aer_num(3)         ! #/cm^3
-   real(r8), intent(out) :: hetraer(3)                  ! BC and Dust mass mean radius [m]
-   real(r8), intent(out) :: awcam(3)                    ! modal added mass [mug m-3]
-   real(r8), intent(out) :: awfacm(3)                   ! (OC+BC)/(OC+BC+SO4)
-   real(r8), intent(out) :: dstcoat(3)                  ! coated fraction
-   real(r8), intent(out) :: na500                  ! #/cm^3 interstitial aerosol number with D>500 nm (#/cm^3)
-   real(r8), intent(out) :: tot_na500              ! #/cm^3 total aerosol number with D>500 nm (#/cm^3)
+   real(r8), intent(out) :: hetraer(hetfrz_aer_nspec)                   ! BC and Dust mass mean radius [m]
+
+   ! local variables
+   real(r8) :: bc_num, dst1_num, dst3_num                ! number concentration [#/cm^3]
+   real(r8), parameter :: aermc_min_threshold  = 1.0e-30_r8
+   real(r8), parameter :: aernum_min_threshold = 1.0e-3_r8
+   real(r8), parameter :: r_bc_prescribed = 0.067e-6_r8
+   real(r8), parameter :: r_dust_a1_prescribed = 0.258e-6_r8
+   real(r8), parameter :: r_dust_a3_prescribed = 1.576e-6_r8
+
+   hetraer = 0.0_r8
+
+   bc_num   = total_interstitial_aer_num(1)
+   dst1_num = total_interstitial_aer_num(2)
+   dst3_num = total_interstitial_aer_num(3)
+
+   ! Initialize hetraer with prescribed radius 
+   hetraer(1) = r_bc_prescribed
+   hetraer(2) = r_dust_a1_prescribed
+   hetraer(3) = r_dust_a3_prescribed
+
+   ! BC
+   if ((aer(bc_accum)+aer(bc_pcarbon))*1.0e-3_r8 > aermc_min_threshold .and. bc_num > aernum_min_threshold) then
+      hetraer(1) = get_aer_radius(specdens_bc, aer(bc_accum)+aer(bc_pcarbon), bc_num*num_cm3_to_m3)
+   endif
+
+   ! fine dust a1
+   if (aer(dst_accum)*1.0e-3_r8 > aermc_min_threshold .and. dst1_num > aernum_min_threshold) then
+      hetraer(2) = get_aer_radius(specdens_dust, aer(dst_accum), dst1_num*num_cm3_to_m3)
+   endif
+
+   ! coarse dust a3
+   if (aer(dst_coarse)*1.0e-3_r8 > aermc_min_threshold .and. dst3_num > aernum_min_threshold) then
+      hetraer(3) = get_aer_radius(specdens_dust, aer(dst_coarse), dst3_num*num_cm3_to_m3)
+   endif
+
+end subroutine calculate_mass_mean_radius
+
+!====================================================================================================
+
+pure function get_aer_radius(specdens, aermc, aernum) result(radius)
+
+   ! Compute radius for an aerosol species given mass and number concentrations
+
+   implicit none
+   real(r8), intent(in) :: specdens    ! density [kg/m^3]
+   real(r8), intent(in) :: aermc       ! aerosol mass concentration [kg/m^3]
+   real(r8), intent(in) :: aernum      ! aerosol number concentration [#/m^3]
+
+   real(r8) :: radius                  ! return value radius [m]
+
+   radius = (3._r8/(4*pi*specdens)*aermc/aernum)**(1._r8/3._r8)
```

Comment:
Please add a note for the C++ port here that this function may already be present in the C++ utility functions.

Reply:
note added